Pass ip, ips, and hostname to Request object

shwei opened the pull request:
#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message and code follows [Code of conduct](https://github.com/fastify/fastify/blob/master/CODE_OF_CONDUCT.md)

Add ip, ips, and hostname in `Request` object. Related to @delvedor's feedback at https://github.com/fastify/fastify/pull/1086#discussion_r211718687. All credits go to @nwoltman . Without his guidance, the work won't be possible. 

[benchmarks](https://gist.github.com/shwei/390c850ebc7a52650de5e33481f5fce3#file-fastify-3dc46f3-tables-comparing-master-vs-pass-ip-to-request-constructor-branches-L1)

nwoltman commented on fastify.js:
```diff
@@ -271,15 +271,11 @@ function build (options) {
-    if (req.ip !== undefined) {
-      req.hostname = req.headers['x-forwarded-host']
-    }
+    return { ip: proxyAddr(req, proxyFn), ips: proxyAddr.all(req, proxyFn), get hostname () { return this.ip !== undefined ? req.headers['x-forwarded-host'] : req.hostname } }
```
`hostname` doesn't need to be a getter. It can be a regular property.

Also, `req.hostname` isn't part of Node's [`IncomingMessage`](https://nodejs.org/dist/latest-v11.x/docs/api/http.html#http_class_http_incomingmessage) object, so `: req.hostname` should be replaced with `: undefined` here.

shwei replied:
I did try it `hostname: this.ip !== undefined ? req.headers['x-forwarded-host'] : undefined ` based on your previous comment, but trust-proxy.test.js tests failed
```
test/trust-proxy.test.js ............................. 5/17 30s
  trust proxy
  not ok Cannot read property 'ip' of undefined
    stack: |
      _handleTrustProxy (fastify.js:274:92)
      Object.routeHandler [as handler] (fastify.js:297:28)
      Router.lookup (node_modules/find-my-way/index.js:338:14)
    at:
      line: 274
      column: 92
      file: fastify.js
      function: _handleTrustProxy
    type: TypeError
    test: trust proxy
    source: >
      return { ip: proxyAddr(req, proxyFn), ips: proxyAddr.all(req, proxyFn),
      hostname: this.ip !== undefined ? req.headers['x-forwarded-host'] : undefined
      }
```

nwoltman replied:
Oh I see what happened. Looks like `ip` should be stored as a variable and then you can replace `this.ip` with just `ip`.

shwei replied:
resolved e446c9f

shwei commented:
> Is this a breaking change?
> We are no longer writing the values in the Node Core request object, but in ours.

Hi @delvedor: Do you mean this particular commit a breaking change or this pull request?  

The call to `handleTrustProxy(req)` assigns req.ip to some value depending on the proxy option. Isn't the req object in this case a Node Core one? https://github.com/fastify/fastify/blob/master/fastify.js#L303

Ethan-Arrowood commented on fastify.js:
```diff
@@ -271,15 +271,12 @@ function build (options) {
-      req.hostname = req.headers['x-forwarded-host']
-    }
+    var ip = proxyAddr(req, proxyFn)
+    return { ip: ip, ips: proxyAddr.all(req, proxyFn), hostname: ip !== undefined ? req.headers['x-forwarded-host'] : undefined }
```
Small nit-pick; for code readability can you new line each `key: value` pair?

mcollina replied:
I don't understand this change.

Ethan-Arrowood replied:
From:
```js
return { ip: ip, ips: proxyAddr.all(req, proxyFn), hostname: ip !== undefined ? req.headers['x-forwarded-host'] : undefined }
```
to:
```js
return { 
  ip: ip, 
  ips: proxyAddr.all(req, proxyFn), 
  hostname: ip !== undefined ? req.headers['x-forwarded-host'] : undefined 
}
```
With both a method call and a ternary expression, this code seems very cluttered all in one line. Totally just a preference and can be ignored

shwei replied:
Refactored the function that's determining `ip`, `ips`, and `hostname` in the trust-proxy scenario. [43657f8](https://github.com/fastify/fastify/commit/43657f8357db5ebe896d9909dd1d1b7ad5d3fd0e)

Ethan-Arrowood commented on lib/request.js:
```diff
@@ -1,22 +1,28 @@
 'use strict'
 
-function Request (params, req, query, headers, log) {
+function Request (params, req, query, headers, log, ip, ips, hostname) {
```
I think this change will require an update in our types file as well

mcollina replied:
Not the constructor, but the new properties in `Request`  needs to be added.

Ethan-Arrowood replied:
I believe the types need to be added in this type declaration block: https://github.com/fastify/fastify/blob/master/fastify.d.ts#L137

shwei replied:
> I believe the types need to be added in this type declaration block: https://github.com/fastify/fastify/blob/master/fastify.d.ts#L137

Thank you, @Ethan-Arrowood. Added the property, ips in Request declaration block. b542fb8

mcollina commented on fastify.js:
```diff
@@ -271,15 +271,12 @@ function build (options) {
 
   function _ipAsRemoteAddress (req) {
-    req.ip = req.connection.remoteAddress
+    return { ip: req.connection.remoteAddress, ips: undefined, hostname: undefined }
```
Can you refactor this code so it does not allocate any new object? This is a super-hot path.

shwei replied:
Thank you, @mcollina. Refactored the logic so that it does not create new object.  43657f8

nwoltman commented:
@delvedor 

> Is this a breaking change?
> We are no longer writing the values in the Node Core request object, but in ours

Yes, removing those values makes this a breaking change. Another option would be to add the values to Fastify's `Request` object now, and remove them from the Node core `req` object for the next major.

Ethan-Arrowood commented:
Changes LGTM ✨

mcollina commented on fastify.d.ts:
```diff
@@ -146,6 +146,7 @@ declare namespace fastify {
     id: any
 
     ip: string
+    ips: any
```
Can this really be a `any`? What's the output of the `proxyAddr.all` function?

shwei replied:
proxyAddr.all returns an array strings. Updated the type to reflect that. efef8cb

mcollina commented on fastify.js:
```diff
@@ -298,16 +285,27 @@ function build (options) {
+    var ip = req.connection.remoteAddress
+    var ips
+
+    if (options.trustProxy) {
```
Can you please make this a `const` in the outer scope?

shwei replied:
Done and there are references to `options.trustProxy` in getTrustProxyFn function also. Updated them accordingly. fe3b111

mcollina commented on fastify.js:
```diff
@@ -298,16 +285,27 @@ function build (options) {
+
+    var hostname = req.headers['host']
+    var ip = req.connection.remoteAddress
+    var ips
```
Can you add a test that checks that `ips`  is `undefined`?

shwei replied:
If you meant creating a test case for undefined `ips`, currently there are some use cases being tested in [trust-proxy.test.js](https://github.com/fastify/fastify/blob/master/test/trust-proxy.test.js#L39) Is this what you meant or am I way off? Thanks!

shwei commented:
master branch:
```>$ git checkout master
Switched to branch 'master'
Your branch is ahead of 'origin/master' by 6 commits.
  (use "git push" to publish your local commits)
>$ node examples/example.js 
server listening on 3000
>$ autocannon -c 100 -d 40 -p 10 localhost:3000
Running 40s test @ http://localhost:3000
100 connections with 10 pipelining factor

┌─────────┬──────┬──────┬───────┬───────┬─────────┬──────────┬───────────┐
│ Stat    │ 2.5% │ 50%  │ 97.5% │ 99%   │ Avg     │ Stdev    │ Max       │
├─────────┼──────┼──────┼───────┼───────┼─────────┼──────────┼───────────┤
│ Latency │ 0 ms │ 0 ms │ 47 ms │ 54 ms │ 4.31 ms │ 13.76 ms │ 400.59 ms │
└─────────┴──────┴──────┴───────┴───────┴─────────┴──────────┴───────────┘
┌───────────┬─────────┬─────────┬─────────┬─────────┬─────────┬─────────┬─────────┐
│ Stat      │ 1%      │ 2.5%    │ 50%     │ 97.5%   │ Avg     │ Stdev   │ Min     │
├───────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┤
│ Req/Sec   │ 11895   │ 11895   │ 22175   │ 26943   │ 22765.1 │ 3482.72 │ 11890   │
├───────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┤
│ Bytes/Sec │ 1.95 MB │ 1.95 MB │ 3.64 MB │ 4.42 MB │ 3.73 MB │ 571 kB  │ 1.95 MB │
└───────────┴─────────┴─────────┴─────────┴─────────┴─────────┴─────────┴─────────┘

Req/Bytes counts sampled once per second.

911k requests in 40.07s, 149 MB read
```

pass-ip-to-request-constructor:
```
>$ git checkout pass-ip-to-request-constructor
Switched to branch 'pass-ip-to-request-constructor'
Your branch is ahead of 'origin/pass-ip-to-request-constructor' by 2 commits.
  (use "git push" to publish your local commits)
>$ node examples/example.js 
server listening on 3000
>$ autocannon -c 100 -d 40 -p 10 localhost:3000
Running 40s test @ http://localhost:3000
100 connections with 10 pipelining factor

┌─────────┬──────┬──────┬───────┬───────┬─────────┬──────────┬───────────┐
│ Stat    │ 2.5% │ 50%  │ 97.5% │ 99%   │ Avg     │ Stdev    │ Max       │
├─────────┼──────┼──────┼───────┼───────┼─────────┼──────────┼───────────┤
│ Latency │ 0 ms │ 0 ms │ 47 ms │ 54 ms │ 4.27 ms │ 13.28 ms │ 409.13 ms │
└─────────┴──────┴──────┴───────┴───────┴─────────┴──────────┴───────────┘
┌───────────┬─────────┬─────────┬─────────┬─────────┬─────────┬─────────┬─────────┐
│ Stat      │ 1%      │ 2.5%    │ 50%     │ 97.5%   │ Avg     │ Stdev   │ Min     │
├───────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┤
│ Req/Sec   │ 12271   │ 12271   │ 23583   │ 26463   │ 23005.3 │ 2918.96 │ 12270   │
├───────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┤
│ Bytes/Sec │ 2.01 MB │ 2.01 MB │ 3.87 MB │ 4.34 MB │ 3.77 MB │ 478 kB  │ 2.01 MB │
└───────────┴─────────┴─────────┴─────────┴─────────┴─────────┴─────────┴─────────┘

Req/Bytes counts sampled once per second.

920k requests in 40.07s, 151 MB read
```

Per @mcollina 's comment:

> Can you run a benchmark, like so:
> 
> ```
> node examples/example
> ```
> ```
> autocannon -c 100 -d 40 -p 10 localhost:3000
> ```
> And post the results, before and after this PR?

nwoltman commented on fastify.js:
```diff
@@ -298,16 +286,27 @@ function build (options) {
+      ip = proxyAddr(req, proxyFn)
+      ips = proxyAddr.all(req, proxyFn)
+      if (ip !== undefined) {
+        hostname = req.headers['x-forwarded-host']
```
If the `x-forwarded-host` header is not available, the `host` header should be used as a fallback.
A test should be added for this case.

shwei replied:
Are you referring to something like ```
 if (ip !== undefined && req.headers['x-forwarded-host']) {
        hostname = req.headers['x-forwarded-host']
``` ?

And do you know what does this mean? Not sure how to do this...
> Can you please make this change non semver major? The properties must still be on the raw objects.

nwoltman replied:
> Are you referring to something like
>
> ```
> if (ip !== undefined && req.headers['x-forwarded-host']) {
>        hostname = req.headers['x-forwarded-host']
> ```
> ?

Yes, that looks good 👍

> And do you know what does this mean?

To "make this change non semver major" means to make changes to this PR so that it doesn't break any existing code (so the PR can be accepted *without* needing to increase the [semver](https://semver.org/) major number).

To do this, you just need to add the `ip`, `ips`, and `hostname` properties back to the Node `req` object. That way, any code that currently relies on those properties won't break after this PR is merged. The properties can be removed in a separate PR (which will get labelled as "semver major").

shwei replied:
@nwoltman Thanks! A test is added in 3ce1c00. I also added ip, ips, and hostname back in 933d4d3, but the whole reason for this PR is to make fastify functional on Google Cloud Function. https://github.com/fastify/fastify/issues/946#issuecomment-463037727. So is it in your roadmap that the removal can only happen at 3.x.x? And if so, when will that be?

nwoltman replied:
Yes, the removal will happen only for v3.0.0. There currently isn't a plan for when v3 will be released, but as the [LTS schedule](https://github.com/fastify/fastify/blob/master/docs/LTS.md) states, v2 will be supported for at least 6 months, so v3 won't be realeased for at least that long.

It's an unfortunate situation, but all we can do is wait. In the meantime, at least you can rely on [this workaround](https://github.com/fastify/fastify/issues/946#issuecomment-469540097).

mcollina commented on test/trust-proxy.test.js:
```diff
@@ -18,15 +18,15 @@ const sgetForwardedRequest = (app, forHeader, path) => {
   }
   if (options.ips) {
-    t.deepEqual(req.raw.ips, options.ips, 'gets ips from x-forwarder-for')
+    t.deepEqual(req.ips, options.ips, 'gets ips from x-forwarder-for')
```
Can you also keep the previous assertions?

shwei replied:
Done c680bee

mcollina commented on docs/Server.md:
````diff
@@ -262,6 +262,26 @@ const fastify = require('fastify')({
+```
+
++ Default: `false`
++ `true/false`: add properties `ip`, `ips`, `hostname`, and `log` to Node's request object and only `log` to Node's response object  (`true`) or leave Node's request and response objects alone (`false`).
````
Maybe the name is better to be something like “modifyCoreObjects” instead? We are also avoiding changing res, and the option esplicitly mention req.

mcollina commented on fastify.js:
```diff
@@ -874,13 +884,18 @@ function build (options) {
+    var logger = log.child({ reqId: req.id })
+    if (optAddToNodeReq) {
+      req.log = res.log = logger
+    }
```
We might want to add ip here as well.
Would that be useful?

shwei replied:
Are you thinking about something like this?
```javascript 
    if (optAddToNodeReq) {
      req.ip = req.connection.remoteAddress
      req.log = res.log = logger
    }
```

shwei commented:
Before adding modifyCoreObjects option:
```
>$ autocannon -c 100 -d 40 -p 10 localhost:3000
Running 40s test @ http://localhost:3000
100 connections with 10 pipelining factor

┌─────────┬──────┬──────┬───────┬───────┬─────────┬─────────┬───────────┐
│ Stat    │ 2.5% │ 50%  │ 97.5% │ 99%   │ Avg     │ Stdev   │ Max       │
├─────────┼──────┼──────┼───────┼───────┼─────────┼─────────┼───────────┤
│ Latency │ 0 ms │ 0 ms │ 56 ms │ 66 ms │ 5.22 ms │ 16.7 ms │ 442.56 ms │
└─────────┴──────┴──────┴───────┴───────┴─────────┴─────────┴───────────┘
┌───────────┬─────────┬─────────┬─────────┬─────────┬─────────┬─────────┬─────────┐
│ Stat      │ 1%      │ 2.5%    │ 50%     │ 97.5%   │ Avg     │ Stdev   │ Min     │
├───────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┤
│ Req/Sec   │ 9327    │ 9327    │ 19135   │ 22191   │ 18827.5 │ 3020.35 │ 9322    │
├───────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┤
│ Bytes/Sec │ 1.53 MB │ 1.53 MB │ 3.14 MB │ 3.64 MB │ 3.09 MB │ 495 kB  │ 1.53 MB │
└───────────┴─────────┴─────────┴─────────┴─────────┴─────────┴─────────┴─────────┘

Req/Bytes counts sampled once per second.

753k requests in 40.07s, 124 MB read
```

After the new option:
```
>$ autocannon -c 100 -d 40 -p 10 localhost:3000
Running 40s test @ http://localhost:3000
100 connections with 10 pipelining factor

┌─────────┬──────┬──────┬───────┬───────┬─────────┬───────┬───────────┐
│ Stat    │ 2.5% │ 50%  │ 97.5% │ 99%   │ Avg     │ Stdev │ Max       │
├─────────┼──────┼──────┼───────┼───────┼─────────┼───────┼───────────┤
│ Latency │ 0 ms │ 0 ms │ 53 ms │ 61 ms │ 4.83 ms │ 15 ms │ 439.34 ms │
└─────────┴──────┴──────┴───────┴───────┴─────────┴───────┴───────────┘
┌───────────┬─────────┬─────────┬─────────┬─────────┬─────────┬─────────┬─────────┐
│ Stat      │ 1%      │ 2.5%    │ 50%     │ 97.5%   │ Avg     │ Stdev   │ Min     │
├───────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┤
│ Req/Sec   │ 13951   │ 13951   │ 20895   │ 23103   │ 20343.8 │ 2162.37 │ 13950   │
├───────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┤
│ Bytes/Sec │ 2.29 MB │ 2.29 MB │ 3.43 MB │ 3.79 MB │ 3.34 MB │ 354 kB  │ 2.29 MB │
└───────────┴─────────┴─────────┴─────────┴─────────┴─────────┴─────────┴─────────┘

Req/Bytes counts sampled once per second.

814k requests in 40.07s, 133 MB read
```

When `modifyCoreObjects = true`:
```
>$ autocannon -c 100 -d 40 -p 10 localhost:3000
Running 40s test @ http://localhost:3000
100 connections with 10 pipelining factor

┌─────────┬──────┬──────┬───────┬───────┬─────────┬──────────┬───────────┐
│ Stat    │ 2.5% │ 50%  │ 97.5% │ 99%   │ Avg     │ Stdev    │ Max       │
├─────────┼──────┼──────┼───────┼───────┼─────────┼──────────┼───────────┤
│ Latency │ 0 ms │ 0 ms │ 55 ms │ 63 ms │ 5.05 ms │ 15.86 ms │ 432.15 ms │
└─────────┴──────┴──────┴───────┴───────┴─────────┴──────────┴───────────┘
┌───────────┬─────────┬─────────┬─────────┬─────────┬─────────┬─────────┬─────────┐
│ Stat      │ 1%      │ 2.5%    │ 50%     │ 97.5%   │ Avg     │ Stdev   │ Min     │
├───────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┤
│ Req/Sec   │ 11295   │ 11295   │ 20031   │ 24015   │ 19452.6 │ 2854.94 │ 11295   │
├───────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┤
│ Bytes/Sec │ 1.85 MB │ 1.85 MB │ 3.28 MB │ 3.94 MB │ 3.19 MB │ 468 kB  │ 1.85 MB │
└───────────┴─────────┴─────────┴─────────┴─────────┴─────────┴─────────┴─────────┘

Req/Bytes counts sampled once per second.

778k requests in 40.08s, 128 MB read
```
> Can you run a quick autocannon benchmark against the example before & after this change with the option on and off?
> 
> Also, the typescript def & test should be updated.

nwoltman commented on docs/Server.md:
````diff
@@ -262,6 +262,26 @@ const fastify = require('fastify')({
+})
+```
+
++ Default: `false`
````
@mcollina For backwards compatibility, shouldn't this option default to `true`?

mcollina replied:
Yes actually! Would you have time to update this?

shwei replied:
Changed in 9b6b23f

nwoltman replied:
@shwei I'm not sure where that commit ended up, but I incorporated your changes in #1517.

shwei replied:
Thanks, @nwoltman! By the way, how would this new option get backported?? (Very eager here!)

nwoltman replied:
The master branch is currently still the main branch for Fastify v2, so this is already on track to be released in the next minor version of Fastify 👍
